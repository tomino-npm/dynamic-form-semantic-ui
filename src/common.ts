import { FormElement, DataSet } from '@tomino/dynamic-form';
import { initCss } from '@tomino/toolbelt';

export type SignatureType = {
  comment: string;
  signature: string;
  verifiedState: 'Pending' | 'Verified' | 'Rejected';
  rejected?: boolean;
  uid: string;
  name: string;
  date: Date;
  setValue?(name: string, value: any): any;
};

export type SignatureHandlers = {
  validateForm(): boolean;
  verifySignature(uid: string, signature: string): Promise<string>;
  sign(
    source: string,
    reject: boolean,
    password: string,
    reason: string,
    submit: boolean
  ): Promise<SignatureType>;
  signatureFont(): string;
};

export type FormControlProps = {
  formControl: FormElement;
  owner: DataSet;
  handlers?: { [index: string]: any };
  readOnly: boolean;
};

export const { css, renderCss } = initCss('form-');
