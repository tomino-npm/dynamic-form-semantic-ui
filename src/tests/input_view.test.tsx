import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { create } from './form_query_data';
import { JSONSchema, FormDefinition, FormModel, config } from '@tomino/dynamic-form';
import { TestComponent } from './common';

const schema: JSONSchema = {
  type: 'object',
  properties: {
    name: {
      type: 'string'
    },
    age: {
      type: 'number',
      minimum: 10
    },
    younger: {
      type: 'string',
      expression: `console.log(this); return this['age'] - 20`
    },
    older: {
      type: 'string',
      expression: `this['age'] + 10`
    }
  },
  required: ['name']
};

const formDefinition: FormDefinition = create.form({
  elements: [
    create.formElement({
      row: 0,
      column: 0,
      width: 14,
      control: 'Input',
      controlProps: {
        label: 'Name'
      },
      label: 'Mimo',
      source: 'name'
    }),
    create.formElement({
      row: 1,
      column: 1,
      width: 7,
      control: 'Input',
      source: 'age',
      label: 'Age: ',
      inline: true
    }),
    create.formElement({
      row: 1,
      column: 10,
      width: 2,
      control: 'Input',
      source: 'younger',
      label: 'Younger:',
      inline: true
    }),
    create.formElement({
      row: 0,
      column: 14,
      width: 2,
      control: 'Input',
      source: 'older',
      label: 'Older'
    })
  ]
});

const formData = {
  name: 'Tomas',
  age: 33
};

describe('Input', () => {
  function componentWithData() {
    const form = new FormModel(formDefinition, schema, formData);

    // just another notation
    return <TestComponent form={form} />;
  }

  const setDirty = jest.fn();

  beforeAll(() => {
    config.setDirty = () => setDirty();
  });

  it('renders correctly', () => {
    const component = renderer.create(componentWithData());
    expect(component).toMatchSnapshot();
  });

  it('changes value and all related formulas', () => {
    const component = renderer.create(componentWithData());
    const root = component.root;
    const age = root.findByProps({ name: 'age' });
    age.props.onChange({ currentTarget: { value: '40' } });
    expect(component).toMatchSnapshot();
    expect(setDirty).toBeCalled();
  });

  return { componentWithData };
});
