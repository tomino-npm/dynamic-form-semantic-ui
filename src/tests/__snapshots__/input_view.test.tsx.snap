// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Input changes value and all related formulas 1`] = `
<div
  className="ui segment ui form"
>
  <style>
    
  </style>
  <div
    className="ui form"
  >
    <div
      className="fields"
    >
      <div
        className="fourteen wide field"
      >
        <label
          htmlFor="name"
        >
          Mimo
        </label>
        <div
          className="ui labeled input"
        >
          <div
            className="ui label label"
            onClick={[Function]}
          >
            Name
          </div>
          <input
            name="name"
            onChange={[Function]}
            type="text"
            value="Tomas"
          />
        </div>
      </div>
      <div
        className="two wide field"
      >
        <label
          htmlFor="older"
        >
          Older
        </label>
        <div
          className="ui disabled input"
        >
          <input
            disabled={true}
            name="older"
            onChange={[Function]}
            tabIndex={-1}
            type="text"
            value={50}
          />
        </div>
      </div>
    </div>
    <div
      className="fields"
    >
      <div
        className="one wide field"
      >
         
      </div>
      <div
        className="inline field"
      >
        <label
          htmlFor="age"
        >
          Age: 
        </label>
        <div
          className="ui input"
        >
          <input
            name="age"
            onChange={[Function]}
            type="text"
            value={40}
          />
        </div>
      </div>
      <div
        className="two wide field"
      >
         
      </div>
      <div
        className="inline field"
      >
        <label
          htmlFor="younger"
        >
          Younger:
        </label>
        <div
          className="ui input"
        >
          <input
            name="younger"
            onChange={[Function]}
            type="text"
            value={20}
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Form Input renders correctly 1`] = `
<div
  className="ui segment ui form"
>
  <style>
    
  </style>
  <div
    className="ui form"
  >
    <div
      className="fields"
    >
      <div
        className="fourteen wide field"
      >
        <label
          htmlFor="name"
        >
          Mimo
        </label>
        <div
          className="ui labeled input"
        >
          <div
            className="ui label label"
            onClick={[Function]}
          >
            Name
          </div>
          <input
            name="name"
            onChange={[Function]}
            type="text"
            value="Tomas"
          />
        </div>
      </div>
      <div
        className="two wide field"
      >
        <label
          htmlFor="older"
        >
          Older
        </label>
        <div
          className="ui disabled input"
        >
          <input
            disabled={true}
            name="older"
            onChange={[Function]}
            tabIndex={-1}
            type="text"
            value={43}
          />
        </div>
      </div>
    </div>
    <div
      className="fields"
    >
      <div
        className="one wide field"
      >
         
      </div>
      <div
        className="inline field"
      >
        <label
          htmlFor="age"
        >
          Age: 
        </label>
        <div
          className="ui input"
        >
          <input
            name="age"
            onChange={[Function]}
            type="text"
            value={33}
          />
        </div>
      </div>
      <div
        className="two wide field"
      >
         
      </div>
      <div
        className="inline field"
      >
        <label
          htmlFor="younger"
        >
          Younger:
        </label>
        <div
          className="ui input"
        >
          <input
            name="younger"
            onChange={[Function]}
            type="text"
            value={13}
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
